Use c10 threadpool for GPU to CPU distributed autograd continuations.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42511 Use c10 threadpool for GPU to CPU distributed autograd continuations.**

DistEngine currently only has a single thread to execute GPU to CPU
continuations as part of the backward pass. This would be a significant
performance bottleneck in cases where we have such continuations and would like
to execute these using all CPU cores.

To alleviate this in this PR, we have the single thread in DistEngine only
dequeue work from the global queue, but then hand off execution of that work to
the c10 threadpool where we call "execute_graph_task_until_ready_queue_empty".

For more context please see:
https://github.com/pytorch/pytorch/issues/40255#issuecomment-663298062.

#Closes: https://github.com/pytorch/pytorch/issues/40255

Differential Revision: [D22917579](https://our.internmc.facebook.com/intern/diff/D22917579/)

pritamdamania committed:
Use c10 threadpool for GPU to CPU distributed autograd continuations.

DistEngine currently only has a single thread to execute GPU to CPU
continuations as part of the backward pass. This would be a significant
performance bottleneck in cases where we have such continuations and would like
to execute these using all CPU cores.

To alleviate this in this PR, we have the single thread in DistEngine only
dequeue work from the global queue, but then hand off execution of that work to
the c10 threadpool where we call "execute_graph_task_until_ready_queue_empty".

For more context please see:
https://github.com/pytorch/pytorch/issues/40255#issuecomment-663298062.

Differential Revision: [D22917579](https://our.internmc.facebook.com/intern/diff/D22917579/)

[ghstack-poisoned]

albanD commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -73,30 +73,61 @@ class DistAccumulateGradCaptureHook
+        inputs.add(i, std::move(variables[i]), c10::nullopt, c10::nullopt);
+      }
+      execute_graph_task_until_ready_queue_empty(
+          /*node_task*/ NodeTask(graphTask, graphRoot, std::move(inputs)),
```
Why do you have to "unpack" the given NodeTask that you poped before executing it here?

albanD replied:
But it was moved out of the queue right? So it could be moved to the lambda and then into the new queue?

pritamdamania87 replied:
Yes, it can be moved to the lambda. Although, the lambda is passed to `at::launch` as `std::function<void()> func`. As a result, when the lambda is passed to `at::launch` it copies the lambda and as a result needs to copy the NodeTask (since it is part of the lambda due to the capture) and thats where it fails. If `at::launch` took the function as reference or had the function as a templated reference this wouldn't be a problem.

albanD commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -280,17 +311,17 @@ void DistEngine::computeDependencies(
-      NodeTask(graph_task, std::move(root_to_execute), InputBuffer(0)),
-      incrementOutstandingTasks);
+  auto graph_task = node_task.base_.lock();
+  TORCH_INTERNAL_ASSERT(graph_task);
```
Is this always true?
For the new use case, you keep the graph_task alive in the parent scope so it's fine.
But is it guaranteed in other cases as well? Or is it fine if another thread encounters an error and destroys the graph_task before you get here?

albanD commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -344,8 +375,7 @@ std::shared_ptr<rpc::FutureMessage> DistEngine::runEngineAndAccumulateGradients(
     execute_graph_task_until_ready_queue_empty(
-        /*graph_task*/ graphTask,
-        /*root_to_execute*/ graphRoot,
+        NodeTask(graphTask, graphRoot, InputBuffer(0)),
```
nit: missing comment about which argument it is to be consistent with next line.
Same for all call sites.

pritamdamania committed:
Update on "Use c10 threadpool for GPU to CPU distributed autograd continuations."


DistEngine currently only has a single thread to execute GPU to CPU
continuations as part of the backward pass. This would be a significant
performance bottleneck in cases where we have such continuations and would like
to execute these using all CPU cores.

To alleviate this in this PR, we have the single thread in DistEngine only
dequeue work from the global queue, but then hand off execution of that work to
the c10 threadpool where we call "execute_graph_task_until_ready_queue_empty".

For more context please see:
https://github.com/pytorch/pytorch/issues/40255#issuecomment-663298062.

#Closes: https://github.com/pytorch/pytorch/issues/40255

Differential Revision: [D22917579](https://our.internmc.facebook.com/intern/diff/D22917579/)

[ghstack-poisoned]

pritamdamania87 commented:
@albanD Thanks for reviewing the PR, I addressed your comments. Could you take another look at the PR?

albanD reviewed (commented):
Looks good to me.
Just the small detail about the NodeTask in case we can avoid re-creating it.

pritamdamania committed:
Update on "Use c10 threadpool for GPU to CPU distributed autograd continuations."


DistEngine currently only has a single thread to execute GPU to CPU
continuations as part of the backward pass. This would be a significant
performance bottleneck in cases where we have such continuations and would like
to execute these using all CPU cores.

To alleviate this in this PR, we have the single thread in DistEngine only
dequeue work from the global queue, but then hand off execution of that work to
the c10 threadpool where we call "execute_graph_task_until_ready_queue_empty".

For more context please see:
https://github.com/pytorch/pytorch/issues/40255#issuecomment-663298062.

#Closes: https://github.com/pytorch/pytorch/issues/40255

Differential Revision: [D22917579](https://our.internmc.facebook.com/intern/diff/D22917579/)

[ghstack-poisoned]
